File: gaussjordan.js
function solveGaussJordan() {
    const matrixInput = document.getElementById("matrixInput").value;
    const lines = matrixInput.split('\n');
    const rows = lines.length;
    const cols = lines[0].split(' ').length;

    const matrix = [];
    for (let i = 0; i < rows; i++) {
        const line = lines[i].split(' ');
        matrix.push([]);
        for (let j = 0; j < cols; j++) {
            matrix[i].push(new Fraction(line[j]));
        }
    }

    let outputText = "";

    // Gauss
    let pivot = new Fraction("0");
    for (let k = 0; k < cols; k++) {
        if (matrix[0][k] !== 0 && pivot.num === 0) {
            pivot = matrix[0][k];
        }
        if (pivot.num !== 0) {
            matrix[0][k] /= pivot;
        }
    }

    for (let i = 0; i < cols - 1; i++) {
        for (let j = i + 1; j < rows; j++) {
            if (matrix[i][i].num === 0) {
                continue;
            }
            const ratio = matrix[j][i] / matrix[i][i];
            outputText += `r${j + 1} -${ratio.num >= 0 ? '-' : '+'} ${Math.abs(ratio)} r${i + 1}\n`;
            pivot = new Fraction("0");
            for (let k = i; k < cols; k++) {
                matrix[j][k] = matrix[j][k] - ratio * matrix[i][k];
                if (matrix[j][k] !== 0 && pivot.num === 0) {
                    pivot = matrix[j][k];
                }
                if (pivot.num !== 0) {
                    matrix[j][k] /= pivot;
                }
            }
        }
        outputText += '\n';
    }

    // Output
    outputText += "Result:\n";
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            outputText += matrix[i][j].toString() + ' ';
        }
        outputText += '\n';
    }

    document.getElementById("output").textContent = outputText;
}

class Fraction {
    constructor(frac) {
        this.num = parseInt(frac.split("/")[0]);
        this.den = frac.includes("/") ? parseInt(frac.split("/")[1]) : 1;
        this.reduce();
    }

    reduce() {
        const commonDivisor = this.gcd(this.den, this.num);
        this.den = this.den / commonDivisor;
        this.num = this.num / commonDivisor;
        if (this.den < 0) {
            this.den *= -1;
            this.num *= -1;
        }
    }

    gcd(a, b) {
        a = Math.abs(a);
        b = Math.abs(b);
        while (b > 0) {
            const c = a % b;
            a = b;
            b = c;
        }
        return a;
    }

    add(other) {
        const res = new Fraction("1");
        const lcm = (this.den * other.den) / this.gcd(this.den, other.den);
        res.den = lcm;
        res.num = (this.num * (lcm / this.den)) + (other.num * (lcm / other.den));
        res.reduce();
        return res;
    }

    subtract(other) {
        const res = new Fraction("1");
        const lcm = (this.den * other.den) / this.gcd(this.den, other.den);
        res.den = lcm;
        res.num = (this.num * (lcm / this.den)) - (other.num * (lcm / other.den));
        res.reduce();
        return res;
    }

    multiply(other) {
        const res = new Fraction("1");
        res.den = this.den * other.den;
        res.num = this.num * other.num;
        res.reduce();
        return res;
    }

    divide(other) {
        const res = new Fraction("1");
        res.den = this.den * other.num;
        res.num = this.num * other.den;
        res.reduce();
        return res;
    }

    toString() {
        if (this.den !== 1) {
            return `${this.num}/${this.den}`;
        } else {
            return `${this.num}`;
        }
    }

    abs() {
        const res = new Fraction("1");
        res.num = Math.abs(this.num);
        res.den = this.den;
        return res;
    }
}

File: gaussjordan.py
import numpy as np

def gcd(a, b):
    a = abs(a)
    b = abs(b)
    while(b > 0):
        c = a % b
        a = b
        b = c
    return a

def output():
    widths = np.zeros((cols),dtype=int)
    for i in range(rows):
        for j in range(cols):
            widths[j] = max(widths[j],len(matrix[i][j].__str__()))
    for i in range(rows):
        for j in range(cols):
            print(matrix[i][j].__str__().ljust(widths[j]), end=' ')
        print("")
    print("")

class Fraction:
    num = 1
    den = 1

    def __init__(self, frac):
        self.num = int(frac.split("/")[0])
        if(frac.find("/") != -1):
            self.den = int(frac.split("/")[1])
        else:
            self.den = 1
        self.reduce()

    def reduce(self):
        g = gcd(self.den, self.num)
        self.den = int(self.den / g)
        self.num = int(self.num / g)
        if self.den < 0:
            self.den *= -1
            self.num *= -1

    def __truediv__(self,other):
        res = Fraction("1")
        res.den = self.den * other.num
        res.num = self.num * other.den
        res.reduce()
        return res
        
    def __mul__(self,other):
        res = Fraction("1")
        res.den = self.den * other.den
        res.num = self.num * other.num
        res.reduce()
        return res

    def __add__(self,other):
        res = Fraction("1")
        lcm = int((self.den * other.den) / gcd(self.den, other.den))
        res.den = lcm
        res.num = self.num * (lcm / self.den) + other.num * (lcm / other.den)
        res.reduce()
        return res

    def __sub__(self,other):
        res = Fraction("1")
        lcm = int((self.den * other.den) / gcd(self.den, other.den))
        res.den = lcm
        res.num = self.num * (lcm / self.den) - other.num * (lcm / other.den)
        res.reduce()
        return res
    
    def __str__(self):
        if self.den != 1:
            return str(self.num) + "/" + str(self.den)
        else:
            return str(self.num)

    def __abs__(self):
        res = Fraction("1")
        res.num = abs(self.num)
        res.den = self.den
        return res

rows = 0
cols = 0

rows = int(input("Enter number of rows:"))
cols = int(input("Enter number of columns:"))

matrix = np.zeros((rows,cols), dtype=object)

print("Enter the matrix (line by line):")
for i in range(rows):
    line = input().split()
    for j in range(len(line)):
        matrix[i][j] = Fraction(line[j])

print("Gauss:")

pivot = Fraction("0")
for k in range(0,cols): #First row
    if matrix[0][k] != 0 and pivot.num == 0:
        pivot = matrix[0][k]
    if pivot.num != 0:
        matrix[0][k] /= pivot

for i in range(0,cols - 1):         #Col position / pivot
    for j in range(i + 1,rows): #Row number
        if matrix[i][i].num == 0:
            continue
        ratio = matrix[j][i] / matrix[i][i]
        print("r",j + 1,"-" if ratio.num >= 0 else "+", abs(ratio),"r",i + 1,sep='')
        pivot = Fraction("0")
        for k in range(i,cols): #Element in Row
            matrix[j][k] = matrix[j][k] - ratio * matrix[i][k]
            if matrix[j][k] != 0 and pivot.num == 0:
                pivot = matrix[j][k]
            if pivot.num != 0:
                matrix[j][k] /= pivot

        output()

        
print("Jordan:")
pivots = min(rows, cols)
for i in range(pivots - 1, 0, -1):   #Col position
    for j in range(i - 1, -1, -1):      #Row number 不包括-1
        if j > rows - 1:
            continue
        ratio = matrix[j][i]
        print("r",j + 1,"-" if ratio.num >= 0 else "+", abs(ratio),"r",i + 1,sep='')
        for k in range(cols - 1, i  - 1, -1): #Element in Row
            matrix[j][k] = matrix[j][k] - ratio * matrix[i][k]
        output()
